src/coreclr/pal/src/thread/process.cpp

```diff
@@ -2175,20 +2178,39 @@ PROCBuildCreateDumpCommandLine(
 Function:
   PROCCreateCrashDump
 
-  Creates crash dump of the process. Can be called from the
-  unhandled native exception handler.
+  Creates crash dump of the process. Can be called from the unhandled
+  native exception handler. Allows only one thread to generate the core
+  dump if serialize is true.
 
-(no return value)
+Return:
+  TRUE - succeeds, FALSE - fails
 --*/
 BOOL
 PROCCreateCrashDump(
     std::vector<const char*>& argv,
     LPSTR errorMessageBuffer,
-    INT cbErrorMessageBuffer)
+    INT cbErrorMessageBuffer,
+    bool serialize)
 {
     _ASSERTE(argv.size() > 0);
     _ASSERTE(errorMessageBuffer == nullptr || cbErrorMessageBuffer > 0);
 
+    if (serialize)
+    {
+        size_t currentThreadId = THREADSilentGetCurrentThreadId();
+        size_t previousThreadId = InterlockedCompareExchange(&g_crashingThreadId, currentThreadId, 0);
+        if (previousThreadId != 0)
+        {
+            // Should never reenter or recurse
+            _ASSERTE(previousThreadId != currentThreadId);
+
+            // The first thread generates the crash info and any other threads are blocked
+            poll(NULL, 0, INFTIM);
```

Comment:
Do we need an ifinite loop around this to handle `EINTR` ?

Reply:
(Other places in the repo use `sleep(1)` in a loop for this purpose - e.g. https://github.com/dotnet/runtime/blob/b66a1043e2390bc3d938553f4782dee14f9a6a4a/src/coreclr/pal/src/exception/signal.cpp#L580-L583.)

Reply:
I don't think so.  We pass SA_RESTART when signals are registered so poll() should never return.  We all have been using poll() in this way in various other PAL api's like ExitProcess, TerminateProcess, etc. in process.cpp.

Reply:
Since the handlers get registered with SA_RESTART, this should be OK.

Reply:
We use poll() here https://github.com/dotnet/runtime/blob/b66a1043e2390bc3d938553f4782dee14f9a6a4a/src/coreclr/pal/src/thread/process.cpp#L1197 and https://github.com/dotnet/runtime/blob/b66a1043e2390bc3d938553f4782dee14f9a6a4a/src/coreclr/pal/src/thread/process.cpp#L3225

Reply:
> Since the handlers get registered with SA_RESTART, this should be OK.

You can always have poll failing with `EAGAIN`. It is better to wrap it in infinite the loop in situations like this than trying to impossible crashes. 

> We use poll() here

It would be nice to be consistent. If we like `poll()` better, we should use it everywhere.

Reply:
I guess a 3rd party could isntall a handler without SA_RESTART.